File: unixfs/fs-node-type.go
package unixfs

import (
	"io/fs"
	"os"

	"github.com/pkg/errors"
)

// fsCursorNodeType is a static node type value
type fsCursorNodeType struct {
	// isDir indicates this is a dir
	isDir bool
	// isFile indicates this is a file
	isFile bool
	// isSymlink indicates this is a symlink
	isSymlink bool
}

// NewFSCursorNodeType_Unknown constructs a FSCursorNodeType with no type.
func NewFSCursorNodeType_Unknown() FSCursorNodeType {
	return &fsCursorNodeType{}
}

// NewFSCursorNodeType_File constructs a FSCursorNodeType for a file.
func NewFSCursorNodeType_File() FSCursorNodeType {
	return &fsCursorNodeType{isFile: true}
}

// NewFSCursorNodeType_Dir constructs a FSCursorNodeType for a directory.
func NewFSCursorNodeType_Dir() FSCursorNodeType {
	return &fsCursorNodeType{isDir: true}
}

// NewFSCursorNodeType_Symlink constructs a FSCursorNodeType for a symlink.
func NewFSCursorNodeType_Symlink() FSCursorNodeType {
	return &fsCursorNodeType{isSymlink: true}
}

// NodeTypeToMode converts a fstree node type into a Mode.
func NodeTypeToMode(nodeType FSCursorNodeType, permissions fs.FileMode) fs.FileMode {
	permissions = permissions & fs.ModePerm // filer non-permissions fields
	if nodeType.GetIsSymlink() {
		return permissions | os.ModeSymlink
	}
	if nodeType.GetIsDirectory() {
		return permissions | os.ModeDir
	}
	if nodeType.GetIsFile() {
		// regular file has no mode bits set
		return permissions
	}
	// unknown
	return os.ModeIrregular
}

// ModeToNodeType converts a os.fileMode to a NodeType.
func FileModeToNodeType(mode fs.FileMode) (FSCursorNodeType, error) {
	switch {
	case mode.IsDir():
		return NewFSCursorNodeType_Dir(), nil
	case mode.IsRegular():
		return NewFSCursorNodeType_File(), nil
	case mode&fs.ModeSymlink != 0:
		return NewFSCursorNodeType_Symlink(), nil
	default:
		return nil, errors.Errorf("unsupported mode / node type: %s", mode.String())
	}
}

// GetIsDirectory returns if the node is a directory.
func (f *fsCursorNodeType) GetIsDirectory() bool {
	return f.isDir
}

// GetIsFile returns if the node is a regular file.
func (f *fsCursorNodeType) GetIsFile() bool {
	return f.isFile
}

// GetIsSymlink returns if the node is a symlink.
func (f *fsCursorNodeType) GetIsSymlink() bool {
	return f.isSymlink
}

// _ is a type assertion
var _ FSCursorNodeType = ((*fsCursorNodeType)(nil))

File: db/unixfs/fs-node-type.ts
import type { FSCursorNodeType } from './fs-cursor.js'

// ModeDir is the directory bit in a file mode (matching Go os.ModeDir).
const ModeDir = 0x80000000
// ModeSymlink is the symlink bit in a file mode (matching Go os.ModeSymlink).
const ModeSymlink = 0x08000000
// ModeIrregular is the irregular bit in a file mode (matching Go os.ModeIrregular).
const ModeIrregular = 0x00080000
// ModePerm is the permission bits mask (matching Go fs.ModePerm).
const ModePerm = 0o777
// ModeTypeMask is the type bits mask (matching Go fs.ModeType).
const ModeTypeMask =
  ModeDir |
  ModeSymlink |
  ModeIrregular |
  0x04000000 |
  0x02000000 |
  0x01000000 |
  0x00800000 |
  0x00400000 |
  0x00200000 |
  0x00100000

// fsCursorNodeType is a static node type value.
class StaticFSCursorNodeType implements FSCursorNodeType {
  private isDir: boolean
  private isFile: boolean
  private isSymlinkVal: boolean

  constructor(isDir: boolean, isFile: boolean, isSymlink: boolean) {
    this.isDir = isDir
    this.isFile = isFile
    this.isSymlinkVal = isSymlink
  }

  getIsDirectory(): boolean {
    return this.isDir
  }

  getIsFile(): boolean {
    return this.isFile
  }

  getIsSymlink(): boolean {
    return this.isSymlinkVal
  }
}

// newFSCursorNodeType_Unknown constructs a FSCursorNodeType with no type.
export function newFSCursorNodeType_Unknown(): FSCursorNodeType {
  return new StaticFSCursorNodeType(false, false, false)
}

// newFSCursorNodeType_File constructs a FSCursorNodeType for a file.
export function newFSCursorNodeType_File(): FSCursorNodeType {
  return new StaticFSCursorNodeType(false, true, false)
}

// newFSCursorNodeType_Dir constructs a FSCursorNodeType for a directory.
export function newFSCursorNodeType_Dir(): FSCursorNodeType {
  return new StaticFSCursorNodeType(true, false, false)
}

// newFSCursorNodeType_Symlink constructs a FSCursorNodeType for a symlink.
export function newFSCursorNodeType_Symlink(): FSCursorNodeType {
  return new StaticFSCursorNodeType(false, false, true)
}

// defaultPermissions returns the default permissions set for a filetype.
export function defaultPermissions(nt: FSCursorNodeType): number {
  if (nt.getIsSymlink()) {
    return 0o777
  }
  if (nt.getIsDirectory()) {
    return 0o755
  }
  return 0o644
}

// nodeTypeToMode converts a FSCursorNodeType into a mode value.
export function nodeTypeToMode(
  nodeType: FSCursorNodeType,
  permissions: number,
): number {
  permissions = permissions & ModePerm
  if (nodeType.getIsSymlink()) {
    return permissions | ModeSymlink
  }
  if (nodeType.getIsDirectory()) {
    return permissions | ModeDir
  }
  if (nodeType.getIsFile()) {
    return permissions
  }
  return ModeIrregular
}

// fileModeToNodeType converts a file mode to a FSCursorNodeType.
// Throws if the mode represents an unsupported type.
export function fileModeToNodeType(mode: number): FSCursorNodeType {
  if ((mode & ModeDir) !== 0) {
    return newFSCursorNodeType_Dir()
  }
  if ((mode & ModeTypeMask) === 0) {
    return newFSCursorNodeType_File()
  }
  if ((mode & ModeSymlink) !== 0) {
    return newFSCursorNodeType_Symlink()
  }
  throw new Error('unsupported mode / node type: 0x' + mode.toString(16))
}
